File: python-gdb-activities-20/activity19/gdb/tests/test_account_service.py
# gdb/tests/test_account_service.py
import unittest
from gdb.logging.memory_log_destination import MemoryLogDestination
from gdb.logging.transaction_logger import TransactionLogger
from gdb.service.account_service import AccountService
from gdb.exceptions.account_exception import AccountException
from gdb.exceptions.invalid_pin_exception import InvalidPinException

class TestAccountService(unittest.TestCase):
    def setUp(self):
        # ============================================================
        # 📝 STEP 9: Set Up The Service
        #
        # INSTRUCTIONS:
        #   1. self.dest = MemoryLogDestination()
        #   2. self.logger = TransactionLogger(self.dest)
        #   3. self.service = AccountService(self.logger)
        # ============================================================
        # TODO: wire the in-memory log destination, the logger and the service
        raise NotImplementedError("TODO: Step 9 - set up the AccountService")

    def test_open_and_get_account(self):
        # ============================================================
        # 📝 STEP 10: Open And Look Up An Account
        #
        # INSTRUCTIONS:
        #   1. Open a Savings account for "Rajesh Sharma" (30 yrs, Rs. 20,000, tenure 2) with open_account().
        #   2. Assert its account number is 1001 (the first number the service hands out).
        #   3. Fetch it again with get_account(1001) and assert the holder name is "Rajesh Sharma".
        # ============================================================
        # TODO: open an account and assert it can be looked up by number
        raise NotImplementedError("TODO: Step 10 - test open_account and get_account")

    def test_deposit(self):
        # ============================================================
        # 📝 STEP 11: Deposit Through The Service
        #
        # INSTRUCTIONS:
        #   1. Open a Savings account (Rs. 20,000) and set PIN 1234.
        #   2. Deposit Rs. 5,000 with service.deposit(1001, 5000.0).
        #   3. Assert the balance is 25000.0 and the transaction history holds 1 record.
        # ============================================================
        # TODO: deposit through the service and assert balance and history
        raise NotImplementedError("TODO: Step 11 - test deposit")

    def test_withdraw(self):
        # ============================================================
        # 📝 STEP 12: Withdraw Through The Service
        #
        # INSTRUCTIONS:
        #   1. Open a Savings account (Rs. 20,000) and set PIN 1234.
        #   2. Withdraw Rs. 5,000 with service.withdraw(1001, 5000.0, 1234).
        #   3. Assert the balance is 15000.0 and the transaction history holds 1 record.
        # ============================================================
        # TODO: withdraw through the service and assert balance and history
        raise NotImplementedError("TODO: Step 12 - test withdraw")

    def test_transfer(self):
        # ============================================================
        # 📝 STEP 13: Transfer Through The Service
        #
        # INSTRUCTIONS:
        #   1. Open a Savings account #1001 (Rs. 50,000, PIN 1234) and a Current account #1002 (Rs. 30,000, PIN 5678).
        #   2. Transfer Rs. 10,000 from 1001 to 1002 with PIN 1234.
        #   3. Assert both balances are now 40000.0 and the history holds 1 record.
        # ============================================================
        # TODO: transfer through the service and assert both balances and the history
        raise NotImplementedError("TODO: Step 13 - test transfer")

    def test_close_account(self):
        # ============================================================
        # 📝 STEP 14: Close An Account
        #
        # INSTRUCTIONS:
        #   1. Open a Savings account and set PIN 1234.
        #   2. Close it with service.close_account(1001, 1234).
        #   3. Assert the account is no longer active (is_active() is False).
        # ============================================================
        # TODO: close an account through the service and assert it is inactive
        raise NotImplementedError("TODO: Step 14 - test close_account")

    def test_error_cases(self):
        # ============================================================
        # 📝 STEP 15: Error Cases
        #
        # INSTRUCTIONS:
        #   1. Assert that depositing into a non-existent account (9999) raises AccountException.
        #   2. Open a Savings account with PIN 1234 and assert that withdrawing with PIN 9999
        #      raises InvalidPinException.
        #
        # HINT: Use `with self.assertRaises(...):` for each case.
        # ============================================================
        # TODO: assert the service rejects unknown accounts and wrong PINs
        raise NotImplementedError("TODO: Step 15 - test error cases")

if __name__ == "__main__":
    unittest.main()

File: java-gdb-activities-20/activity19/src/com/gdb/tests/TestAccountService.java
package com.gdb.tests;

import com.gdb.domain.IAccount;
import com.gdb.domain.Transaction;
import com.gdb.command.TransactionCommand;
import com.gdb.logging.*;
import com.gdb.service.AccountService;
import com.gdb.exceptions.AccountException;
import com.gdb.exceptions.InvalidPinException;
import java.util.List;

public class TestAccountService {
    public static void main(String[] args) throws Exception {
        System.out.println("=".repeat(60));
        System.out.println("  ACTIVITY 19 — ACCOUNT SERVICE DEMO");
        System.out.println("=".repeat(60));

        // ============================================================
        // 📝 STEP 12: Set Up The Service
        //
        // INSTRUCTIONS:
        //   1. LogDestination dest = new MemoryLogDestination();
        //   2. TransactionLogger logger = new TransactionLogger(dest);
        //   3. AccountService service = new AccountService(logger);
        // ============================================================
        // TODO: initialize MemoryLogDestination, TransactionLogger, and AccountService

        // ============================================================
        // 📝 STEP 13: Open Two Accounts
        //
        // INSTRUCTIONS:
        //   1. IAccount john = service.openAccount("SAVINGS", "John Doe", 25, 15000);
        //   2. john.setPin(1234);
        //   3. IAccount jane = service.openAccount("SAVINGS", "Jane Smith", 30, 10000);
        //   4. jane.setPin(5678);
        //   5. Print both accounts using getAccountInfo().
        // ============================================================
        // TODO: open accounts for John and Jane

        // ============================================================
        // 📝 STEP 14: Deposit into John's Account
        //
        // INSTRUCTIONS:
        //   1. Transaction txn = service.deposit(john.getAccountNumber(), 5000);
        //   2. Print txn.
        // ============================================================
        // TODO: execute deposit through service

        // ============================================================
        // 📝 STEP 15: Withdraw from John's Account
        //
        // INSTRUCTIONS:
        //   1. Transaction wtxn = service.withdraw(john.getAccountNumber(), 2000, 1234);
        //   2. Print wtxn.
        // ============================================================
        // TODO: execute withdrawal through service

        // ============================================================
        // 📝 STEP 16: Transfer from John to Jane
        //
        // INSTRUCTIONS:
        //   1. Transaction ttxn = service.transfer(john.getAccountNumber(), jane.getAccountNumber(), 1000, 1234);
        //   2. Print ttxn.
        // ============================================================
        // TODO: execute transfer through service

        // ============================================================
        // 📝 STEP 17: Print Final Balances
        //
        // INSTRUCTIONS:
        //   Print john.getBalance() and jane.getBalance().
        // ============================================================
        // TODO: print updated balances

        // ============================================================
        // 📝 STEP 18: Print Transaction History
        //
        // INSTRUCTIONS:
        //   1. List<TransactionCommand> history = service.getTransactionHistory();
        //   2. Print size and each command's transaction.
        // ============================================================
        // TODO: print full transaction history

        // ============================================================
        // 📝 STEP 19: Test Error Cases
        //
        // INSTRUCTIONS:
        //   1. Try deposit to non-existent account 9999 -> catch AccountException.
        //   2. Try withdraw with wrong PIN -> catch InvalidPinException.
        // ============================================================
        // TODO: verify exception handling for invalid operations
    }
}
